src/code_0807CC3C.c

```diff
@@ -0,0 +1,543 @@
+    return ret;
+}
+
+u32 sub_0807D128(const Thing* thing) {
+    u32 ret;
+    switch (thing->unk_3) {
+        case 'MCZ3':
+            if (thing->unk_1 + thing->unk_2 == 0x10000) {
+                ret = 2;
+            } else {
+                ret = 0;
+            }
+            break;
+        case 'FleD':
+        case 'TINI':
+            ret = 0;
+            if ((thing->unk_1 & thing->unk_2) == 0xffff) {
+                ret = 1;
+            }
+            break;
+        default:
+            ret = 0;
+            break;
+    }
+    return ret;
+}
+
+u32 sub_0807D184(u32 param_1, const char* param_2) {
+    u32 uVar1;
+
+    uVar1 = sub_0807D20C(param_1, param_2, 8);
+    if (uVar1 == 0) {
+        uVar1 = sub_0807D20C(param_1 + 8, param_2, 8);
+    }
+    return uVar1;
+}
+
+u16 sub_0807D1A4(u16* unk_1, u32 unk_2) {
+    u32 uVar1;
+
+    uVar1 = 0;
+    while (unk_2 != 0) {
+        uVar1 = uVar1 + (*unk_1 ^ unk_2);
+        unk_1 = unk_1 + 1;
+        unk_2 = unk_2 - 2;
+    }
+    return uVar1;
+}
+
+struct_0807D1C4* sub_0807D1C4(u32 unk_1) {
+    return &gUnk_0811E4BC[unk_1];
+}
+
+// these three are basically the same and wrong by basically one instruction in the wrong place
+NONMATCH("asm/non_matching/code_0807CC3C/sub_0807D1D8.inc", u32 sub_0807D1D8(u32 unk_1, const char* unk_2, u32 unk_3)) {
+}
+END_NONMATCH
+
+NONMATCH("asm/non_matching/code_0807CC3C/sub_0807D20C.inc", u32 sub_0807D20C(u32 unk_1, const char* unk_2, u32 unk_3)) {
+    unk_3 >>= 3;
+    unk_1 >>= 3;
+    while (unk_3-- > 0) {
+        if (sub_080B18DC(unk_1, unk_2)) {
+            sub_080B18DC(unk_1, gUnk_0811E4B4);
+            return 0;
+        }
+        unk_1++;
+        unk_2 += 8;
+    }
+    return 1;
+}
+END_NONMATCH
+
+NONMATCH("asm/non_matching/code_0807CC3C/sub_0807D24C.inc", u32 sub_0807D24C(u32 unk_1, const char* unk_2, u32 unk_3)) {
+    unk_3 >>= 3;
+    unk_1 >>= 3;
+    while (unk_3-- > 0) {
+        if (sub_080B180C(unk_1, unk_2))
+            return 0;
+        unk_1++;
+        unk_2 += 8;
+    }
+    return 1;
+}
+END_NONMATCH
+
+// these three functions use gRoomControls, maybe once that is understood better, these can be decompiled easier
+NONMATCH("asm/non_matching/code_0807CC3C/sub_0807D280.inc", void sub_0807D280(u32 unk_1, u32 unk_2)) {
+}
+END_NONMATCH
+
+NONMATCH("asm/non_matching/code_0807CC3C/sub_0807D46C.inc", void sub_0807D46C(u32 unk_1, u32 unk_2)) {
+}
+END_NONMATCH
+
+NONMATCH("asm/non_matching/code_0807CC3C/sub_0807D6D8.inc", void sub_0807D6D8(u32 unk_1, u32 unk_2)) {
+}
+END_NONMATCH
+
+void sub_0807DA70(void) {
+    _DmaZero(&gUnk_02033280, sizeof(gUnk_02033280));
+    _DmaZero(&gScriptExecutionContextArray, sizeof(gScriptExecutionContextArray));
+    _DmaZero(&gPlayerScriptExecutionContext, sizeof(gPlayerScriptExecutionContext));
+    gUnk_02033280.unk_08 = 8;
+}
+
+ScriptExecutionContext* CreateScriptExecutionContext(void) {
+    ScriptExecutionContext* puVar1;
+
+    puVar1 = gScriptExecutionContextArray;
+    do {
```

Comment:
This can probably be rewritten as a for loop.

Reply:
I tried again, but I can't get it to match with for